src/Containers/macro.jl

```diff
@@ -29,32 +17,65 @@ function _reorder_parameters(args)
 end
 
 """
-    parse_macro_arguments(error_fn::Function, args)
+    parse_macro_arguments(
+        error_fn::Function,
+        args;
+        valid_kwargs::Union{Nothing,Vector{Symbol}} = nothing,
+        num_positional_args::Union{Nothing,Int,UnitRange{Int}} = nothing,
+    )
 
 Returns a `Tuple{Vector{Any},Dict{Symbol,Any}}` containing the ordered
 positional arguments and a dictionary mapping the keyword arguments.
 
 This specially handles the distinction of `@foo(key = value)` and
 `@foo(; key = value)` in macros.
 
-Throws an error if mulitple keyword arguments are passed with the same name.
+An error is thrown if multiple keyword arguments are passed with the same key.
+
+If `valid_kwargs` is a `Vector{Symbol}`, an error is thrown if a keyword is not
+in `valid_kwargs`.
+
+If `num_positional_args` is not nothing, an error is thrown if the number of
+positional arguments is not in `num_positional_args`.
 """
-function parse_macro_arguments(error_fn::Function, args)
-    pos_args, kw_args = Any[], Dict{Symbol,Any}()
+function parse_macro_arguments(
+    error_fn::Function,
+    args;
+    valid_kwargs::Union{Nothing,Vector{Symbol}} = nothing,
+    num_positional_args::Union{Nothing,Int,UnitRange{Int}} = nothing,
+)
+    pos_args, kwargs = Any[], Dict{Symbol,Any}()
     for arg in _reorder_parameters(args)
```

Comment:
This gives a `BoundsError` if `args` is empty because `_reorder_parameters` assumes there is at least 1 argument. Thus something `@variable()` gives:
```julia
ERROR: LoadError: BoundsError: attempt to access Tuple{} at index [1]
Stacktrace:
 [1] getindex(t::Tuple, i::Int64)
   @ Base .\tuple.jl:29
 [2] _reorder_parameters(args::Tuple{})
   @ JuMP.Containers ~\.julia\packages\JuMP\IqGDd\src\Containers\macro.jl:7
 [3] parse_macro_arguments(error_fn::JuMP.Containers.var"#error_fn#98"{String}, args::Tuple{}; valid_kwargs::Nothing, num_positional_args::Nothing)
   @ JuMP.Containers ~\.julia\packages\JuMP\IqGDd\src\Containers\macro.jl:48
 [4] parse_macro_arguments(error_fn::Function, args::Tuple{})
   @ JuMP.Containers ~\.julia\packages\JuMP\IqGDd\src\Containers\macro.jl:41
 [5] var"@variable"(__source__::LineNumberNode, __module__::Module, input_args::Vararg{Any})
   @ JuMP ~\.julia\packages\JuMP\IqGDd\src\macros\@variable.jl:144
in expression starting at REPL[16]:1
```

Reply:
I would like `_reorder_parameters` to simply return `args` if it is empty.

Reply:
Done